packages/backend-common/src/reading/AwsS3UrlReader.ts

```diff
@@ -156,39 +157,95 @@ export class AwsS3UrlReader implements UrlReader {
    * If accessKeyId and secretAccessKey are missing, the standard credentials provider chain will be used:
    * https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html
    */
-  private static buildCredentials(
+  private static buildStaticCredentials(
+    accessKeyId: string,
+    secretAccessKey: string,
+  ): AwsCredentialIdentityProvider {
+    return async () => {
+      return Promise.resolve({
+        accessKeyId,
+        secretAccessKey,
+      });
+    };
+  }
+
+  private static async buildCredentials(
+    credsManager: AwsCredentialsManager,
+    region: string,
     integration?: AwsS3Integration,
-  ): aws.Credentials | CredentialsOptions | undefined {
+  ): Promise<AwsCredentialIdentityProvider> {
+    // Fall back to the default credential chain if neither account ID
+    // nor explicit credentials are provided
     if (!integration) {
-      return undefined;
+      return (await credsManager.getCredentialProvider()).sdkCredentialProvider;
     }
 
+    // Pull credentials from the techdocs config section (deprecated)
     const accessKeyId = integration.config.accessKeyId;
     const secretAccessKey = integration.config.secretAccessKey;
-    let explicitCredentials: aws.Credentials | undefined;
-
+    let explicitCredentials: AwsCredentialIdentityProvider;
     if (accessKeyId && secretAccessKey) {
-      explicitCredentials = new aws.Credentials({
+      explicitCredentials = AwsS3UrlReader.buildStaticCredentials(
         accessKeyId,
         secretAccessKey,
-      });
+      );
+    } else {
+      explicitCredentials = (await credsManager.getCredentialProvider())
+        .sdkCredentialProvider;
     }
 
     const roleArn = integration.config.roleArn;
     if (roleArn) {
-      return new aws.ChainableTemporaryCredentials({
+      return fromTemporaryCredentials({
         masterCredentials: explicitCredentials,
         params: {
           RoleSessionName: 'backstage-aws-s3-url-reader',
           RoleArn: roleArn,
           ExternalId: integration.config.externalId,
         },
+        clientConfig: { region },
       });
     }
 
     return explicitCredentials;
   }
 
+  async buildS3Client(
+    defaultConfig: Config,
+    region: string,
+    integration: AwsS3Integration,
+  ): Promise<S3Client> {
+    const credsManager = DefaultAwsCredentialsManager.fromConfig(defaultConfig);
```

Comment:
`DefaultAwsCredentialsManager` caches instantiated credential providers - that seems to be one of its main jobs. Am I reading this correctly that we're making new credentials  and clients for every request made? Is that OK?

Reply:
I modified the code to create the `DefaultAwsCredentialsManager` when building the reader. The `S3Client` requires a region, and I'm assuming it can change depending on the requested URL, so right now, the reader is creating a new client for every request. Another approach could be to create a new config with a default region and build a default S3Client and only create a new client if the region changes; however, that would involve changing the config format.